File: MATLAB/Method/Statistical_methods/Time_domain/PR/PR.m
function [Pulse_shape_discrimination_factor] = PR(Pulse_signal)

%--------------------------------------------------------------------------
% Pattern Recognition (PR) method for pulse shape discrimination.
%
% Discriminates particle types by computing the angular separation between
% pulse signals and a reference signal in vector space. Uses the geometric
% relationship between signals to capture shape differences.
%
% Input:
%   Pulse_signal: Matrix of pulse signals to analyze [n_signals × signal_length]
%
% Output:
%   Pulse_shape_discrimination_factor: Angular separation from reference (rad)
%                                    [n_signals × 1]
%
% Algorithm:
%   1. Select reference signal (second signal in matrix)
%   2. For each signal:
%      - Align with reference at peak position
%      - Compute angle θ between vectors:
%        θ = arccos((x·r)/(||x||·||r||))
%      where x is signal vector, r is reference vector
%
% Mathematical basis:
%   - Vectors: Signal amplitudes after peak
%   - Similarity: Cosine of angle between vectors
%   - Range: θ ∈ [0, π] radians
%
% Note:
%   - Requires at least 2 signals (one for reference)
%   - Zero-norm vectors will cause errors
%   - Cosine is clamped to [-1,1] for numerical stability
%
% Reference:
% Takaku, D., T. Oishi, and M. Baba. "Development of neutron-gamma
% discrimination technique using pattern-recognition method with digital
% signal processing." Prog. Nucl. Sci. Technol 1 (2011): 210-213.
%--------------------------------------------------------------------------

% Initialize output array
Num_signal_row = size(Pulse_signal, 1);
Pulse_shape_discrimination_factor = zeros(Num_signal_row, 1);

% Validate input size
if Num_signal_row < 2
    error('At least 2 signals required (1 for analysis, 1 for reference).');
end

% Select reference signal (second row of input matrix)
Reference_signal_original = Pulse_signal(2, :);

% Process signals in parallel
parfor i = 1:Num_signal_row
    % Get current signal
    Pulse_signal_vector = Pulse_signal(i, :);
    
    % Align signals at peak
    [~, Max_position] = max(Pulse_signal_vector);
    Pulse_signal_vector = Pulse_signal_vector(Max_position:end);  % Post-peak region
    Reference_signal = Reference_signal_original(Max_position:end);  % Aligned reference

    % Compute vector operations
    Scalar_product = dot(Pulse_signal_vector, Reference_signal);  % x·r
    Norm_Reference_signal = norm(Reference_signal);               % ||r||
    Norm_pulse_signal = norm(Pulse_signal_vector);               % ||x||

    % Validate vector norms
    if (Norm_pulse_signal == 0 || Norm_Reference_signal == 0)
        error('Zero-norm vector detected at signal %d. Valid signals required.', i);
    end
    
    % Calculate cosine similarity
    Cos_angle = Scalar_product / (Norm_pulse_signal * Norm_Reference_signal);

    % Ensure numerical stability for arccos
    Cos_angle = max(min(Cos_angle, 1), -1);  % Clamp to [-1,1]
    
    % Compute angular separation
    Pulse_shape_discrimination_factor(i) = abs(acos(Cos_angle));  % θ in radians
end

end

File: Method/Statistical_methods/Time_domain/pr.py
"""
Pattern Recognition (PR) method for pulse shape discrimination.

The method computes discrimination factors by measuring the angle between 
each pulse and a reference pulse in vector space, using their post-peak 
portions for comparison.

Reference:
- Takaku, D., T. Oishi, and M. Baba.
  "Development of neutron-gamma discrimination technique using pattern-recognition method with digital signal processing."
  Prog. Nucl. Sci. Technol 1 (2011): 210-213.
"""

import numpy as np


def get_psd_factor(
        pulse_signal: np.ndarray
) -> np.ndarray:
    """
    Calculate PSD factors using Pattern Recognition.
    
    Args:
        pulse_signal: Input signals array of shape (N, L) where N is the number
                     of pulses and L is the length of each pulse. The second
                     pulse (index 1) is used as the reference
    
    Returns:
        numpy.ndarray: Array of PSD factors for each input pulse, computed as
                      |arccos(cos θ)|, where θ is the angle between the pulse
                      and reference vectors
    
    Raises:
        ValueError: If N < 2 or if any post-peak signal has zero norm
    
    Note:
        Only the portion of each pulse after its peak is used for comparison.
        The cosine angle is computed using the dot product formula:
        cos θ = (a·b)/(|a|·|b|)
    """
    num_signals = pulse_signal.shape[0]
    if num_signals < 2:
        raise ValueError("pulse_signal must have at least two rows for reference selection.")

    # Initialize discrimination factors array
    discrimination_factors = np.zeros(num_signals)

    # Select the reference signal as the second row (index 1)
    reference_signal = pulse_signal[1, :]

    for i, signal in enumerate(pulse_signal):
        # Find the peak position (index of the maximum value)
        max_position = np.argmax(signal)

        # Trim the signal and the reference signal from the peak position onward
        trimmed_signal = signal[max_position:]
        trimmed_reference = reference_signal[max_position:]

        # Compute the scalar product and the norms of the trimmed vectors
        scalar_product = np.dot(trimmed_signal, trimmed_reference)
        norm_signal = np.linalg.norm(trimmed_signal)
        norm_reference = np.linalg.norm(trimmed_reference)

        # Check that neither norm is zero
        if norm_signal == 0 or norm_reference == 0:
            raise ValueError("Norm of one of the vectors is zero, cannot compute discrimination factor.")

        # Calculate the cosine of the angle between the two vectors
        cos_angle = scalar_product / (norm_signal * norm_reference)

        # Clamp the cosine value to the valid range for arccos
        cos_angle = max(min(cos_angle, 1), -1)

        # The discrimination factor is the absolute value of the angle (in radians)
        discrimination_factors[i] = abs(np.arccos(cos_angle))

    return discrimination_factors
